src/shims/sync.rs

```diff
@@ -598,17 +600,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let this = self.eval_context_mut();
 
         let id = rwlock_get_or_create_id(this, rwlock_op)?;
-        let active_thread = this.get_active_thread()?;
+        let active_thread = this.get_active_thread();
 
         if this.rwlock_reader_unlock(id, active_thread) {
             // The thread was a reader.
-            if this.rwlock_is_locked(id) {
+            if this.rwlock_is_locked(id).not() {
```

Comment:
@vakaras there was a crucial negation missing here. I added a testcase as well to ensure that with the negation, it does the right thing.